src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,229 @@
+    """
+
+    if invert_register == "sp_0" or invert_register == "sp_1":
+        if invert_register == "sp_0":
+            register_to_reset = 0
+        elif invert_register == "sp_1":
+            register_to_reset = 1
+
+        new_circ = Circuit(1)
+        for data in circuit.raw["queue"]:
+            theta = data.get("init_kwargs", {}).get("theta", None)
+            if data["_target_qubits"][0] == register_to_reset:
+                if theta is None:
+                    new_circ.add(getattr(gates, data["_class"])(0))
+                else:
+                    new_circ.add(
+                        getattr(gates, data["_class"])(0, theta)
+                    )  # <-- Not covered in tests. Thinking about future proofing when we need to invert a random gate that has theta parameter.
+
+    elif invert_register == "sp_t":
+        new_circ = circuit.copy()
+
+    else:
+        raise NameError(
+            f"{invert_register} not recognized. Input "
+            "sp_0"
+            " to reset qubit 0, "
+            "sp_1"
+            " to reset qubit 1, or "
+            "sp_t"
+            " to reset both qubits."
+        )
+
+    inverse_circuit = new_circ.invert()
+
+    return inverse_circuit
+
+
+def GST_execute_circuit(circuit, k, j, nshots=int(1e4), backend=None):
+    """Executes a circuit used in gate set tomography and processes the
+        measurement outcomes for the Pauli Transfer Matrix notation. The circuit
+        should already have noise models implemented, if any, prior to using this
+        function.
+
+        Args:
+        circuit (:class:`qibo.models.Circuit`): The Qibo circuit to be executed.
+        k (int): The index of the state prepared.
+        j (int): The index of the measurement basis.
+        nshots (int, optional): Number of shots to execute circuit with.
+    Returns:
+        numpy.float: Expectation value given by either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
+    """
+
+    nqubits = circuit.nqubits
+    if nqubits != 1 and nqubits != 2:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
+
+    else:
+        if j == 0:
+            return 1.0
+        else:
+            if backend is None:  # pragma: no cover
+                backend = GlobalBackend()
+
+            result = backend.execute_circuit(circuit, nshots=nshots)
+            observables = [symbols.I, symbols.Z, symbols.Z, symbols.Z]
+            observables_list = list(product(observables, repeat=nqubits))[j]
+            observable = 1
+            for q, obs in enumerate(observables_list):
+                if obs is not symbols.I:
+                    observable *= obs(q)
+            observable = SymbolicHamiltonian(observable, nqubits=nqubits)
+            expectation_val = result.expectation_from_samples(observable)
+            return expectation_val
+
+
+def GST(
```

Comment:
Since this is a function, the naming should follow the Python convention for functions